samples/api/swapchain_recreation/swapchain_recreation.cpp

```diff
@@ -0,0 +1,929 @@
+
+	if (!supports_present)
+	{
+		throw std::runtime_error("Default graphics queue does not support preesnt.");
+	}
+}
+
+void SwapchainRecreation::query_surface_format()
+{
+	uint32_t surface_format_count;
+	vkGetPhysicalDeviceSurfaceFormatsKHR(get_gpu_handle(), get_surface(), &surface_format_count, nullptr);
+	std::vector<VkSurfaceFormatKHR> supported_surface_formats(surface_format_count);
+	vkGetPhysicalDeviceSurfaceFormatsKHR(get_gpu_handle(), get_surface(), &surface_format_count, supported_surface_formats.data());
+
+	// We want to get an SRGB image format that matches our list of preferred format candiates
+	// We initialize to the first supported format, which will be the fallback in case none of the preferred formats is available
+	surface_format             = supported_surface_formats[0];
+	auto preferred_format_list = std::vector<VkFormat>{VK_FORMAT_R8G8B8A8_SRGB, VK_FORMAT_B8G8R8A8_SRGB, VK_FORMAT_A8B8G8R8_SRGB_PACK32};
+
+	for (auto &candidate : supported_surface_formats)
+	{
+		if (std::find(preferred_format_list.begin(), preferred_format_list.end(), candidate.format) != preferred_format_list.end())
+		{
+			surface_format = candidate;
+			break;
+		}
+	}
+}
+
+void SwapchainRecreation::query_present_modes()
+{
+	uint32_t present_mode_count = 0;
+	VK_CHECK(vkGetPhysicalDeviceSurfacePresentModesKHR(get_gpu_handle(), get_surface(), &present_mode_count, nullptr));
+
+	present_modes.resize(present_mode_count);
+	VK_CHECK(vkGetPhysicalDeviceSurfacePresentModesKHR(get_gpu_handle(), get_surface(), &present_mode_count, present_modes.data()));
+
+	adjust_desired_present_mode();
+}
+
+void SwapchainRecreation::adjust_desired_present_mode()
+{
+	// The FIFO present mode is guaranteed to be present.
+	if (desired_present_mode == VK_PRESENT_MODE_FIFO_KHR)
+	{
+		return;
+	}
+
+	// When switching to MAILBOX, fallback to IMMEDIATE if not available and back to FIFO if
+	// neither are available.
+	if (desired_present_mode == VK_PRESENT_MODE_MAILBOX_KHR && std::find(present_modes.begin(), present_modes.end(), desired_present_mode) != present_modes.end())
+	{
+		return;
+	}
+
+	desired_present_mode = VK_PRESENT_MODE_IMMEDIATE_KHR;
+	if (std::find(present_modes.begin(), present_modes.end(), desired_present_mode) == present_modes.end())
+	{
+		LOGW("Neither MAILBOX nor IMMEDIATE are supported, falling back to FIFO");
+		desired_present_mode = VK_PRESENT_MODE_FIFO_KHR;
+	}
+}
+
+void SwapchainRecreation::create_render_pass()
+{
+	VkAttachmentDescription attachment = {0};
+	attachment.format                  = surface_format.format;
+	attachment.samples                 = VK_SAMPLE_COUNT_1_BIT;
+	attachment.loadOp                  = VK_ATTACHMENT_LOAD_OP_CLEAR;
+	attachment.storeOp                 = VK_ATTACHMENT_STORE_OP_STORE;
+	attachment.stencilLoadOp           = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
+	attachment.stencilStoreOp          = VK_ATTACHMENT_STORE_OP_DONT_CARE;
+	attachment.initialLayout           = VK_IMAGE_LAYOUT_UNDEFINED;
```

Comment:
It might look like this hits my favorite obscure sync issue :) You have:
- A render-pass with transition from UNDEFINED
- An acquire semaphore wait in COLOR stage
- No custom external subpass dependencies

The problem is, the UNDEFINED transition can trigger a memory write, and when this happens is defined by the subpass dependencies. The default implicit subpass dependency doesn't 'hook' onto any stage except TOP stage (srcStageMask = TOP), and so there's nothing preventing the UNDEFINED transition from happening ASAP after the TOP stage. The problem then is, the semaphore wait doesn't happen before the COLOR stage -- after TOP -- meaning the transition can happen before the semaphore signals. So we have a race.

Some ways to fix:
- Wait on the acquire semaphore in TOP -- this is suboptimal sync wise, as you have e.g. vertex-shading blocked on the semaphore unnecessarily.
- Set initialLayout to VK_IMAGE_LAYOUT_PRESENT_SRC_KHR so there's no transition generated.
- Add an external subpass-dependency which waits on the COLOR stage, so the semaphore is included in the render-pass's dependencies.

Reply:
Ugh right :roll_eyes: the semaphore wait stage doesn't interact well with the implicit layout transition. I think the best solution is to craft the external subpass-dependency, do you mind drafting a change and I'll incorporate it?

Reply:
This should do the trick I believe:

```
		std::array<VkSubpassDependency, 1> dependencies{};
		dependencies[0].srcSubpass      = VK_SUBPASS_EXTERNAL;
		dependencies[0].dstSubpass      = 0;
		// srcStageMask must match WSI semaphore wait stage (pWaitDstStageMask)
		dependencies[0].srcStageMask    = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
		dependencies[0].dstStageMask    = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
		dependencies[0].srcAccessMask   = 0;
		dependencies[0].dstAccessMask   = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
		dependencies[0].dependencyFlags = 0;

		rp_info.dependencyCount        = static_cast<uint32_t>(dependencies.size());
		rp_info.pDependencies          = dependencies.data();
```

Reply:
Thanks, I put that in.